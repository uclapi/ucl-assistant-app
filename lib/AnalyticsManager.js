import * as Amplitude from 'expo-analytics-amplitude'
import Constants from 'expo-constants'

const AMPLITUDE_API_KEY = Constants.manifest.extra
  && Constants.manifest.extra.amplitude
  && Constants.manifest.extra.amplitude.apiKey
  ? Constants.manifest.extra.amplitude.apiKey
  : null

const initialise = () => Amplitude.initialize(AMPLITUDE_API_KEY)

const setUserId = (userId) => {
  // make sure this is unique
  if (userId) {
    Amplitude.setUserId(userId)
  }
}

const setUserProperties = ({ ...userProperties }) => {
  Amplitude.setUserProperties({ ...userProperties })
}

const clearUserProperties = () => {
  Amplitude.clearUserProperties()
}

const logEvent = (eventName, eventProperties) => {
  if (eventProperties) {
    Amplitude.logEventWithProperties(eventName, eventProperties)
  } else {
    Amplitude.logEvent(eventName)
  }
}

const logScreenView = (screenName) => logEvent(`VIEW_SCREEN_${screenName}`)

const events = {
  NOTIFICATIONS_ENABLE: `NOTIFICATIONS_ENABLE`,
  NOTIFICATIONS_SKIP: `NOTIFICATIONS_SKIP`,
  PUSH_NOTIFICATIONS_REGISTER: `PUSH_NOTIFICATIONS_REGISTER`,
  SETTINGS_GIVE_FEEDBACK: `SETTINGS_GIVE_FEEDBACK`,
  SETTINGS_RATE_APP: `SETTINGS_RATE_APP`,
  SETTINGS_VIEW_FAQS: `SETTINGS_VIEW_FAQS`,
}

const AnalyticsManager = {
  clearUserProperties,
  events,
  initialise,
  logEvent,
  logScreenView,
  setUserId,
  setUserProperties,
}

export default (AMPLITUDE_API_KEY && !__DEV__)
  ? AnalyticsManager
  : (
    new Proxy({}, {
      get() {
        return () => { } // do nothing in development mode
      },
    })
  )
