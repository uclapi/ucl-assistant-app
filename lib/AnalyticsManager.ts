import * as Amplitude from 'expo-analytics-amplitude'
import Constants from 'expo-constants'

import configureStore from "../configureStore"

const { store } = configureStore

const AMPLITUDE_API_KEY = Constants.manifest.extra
  && Constants.manifest.extra.amplitude
  && Constants.manifest.extra.amplitude.apiKey
  ? Constants.manifest.extra.amplitude.apiKey
  : null

const initialise = (): Promise<void> => Amplitude.initializeAsync(AMPLITUDE_API_KEY)

const shouldTrackAnalytics = (): boolean => (
  AMPLITUDE_API_KEY
  && !__DEV__
  && store.getState().user.settings.shouldTrackAnalytics
)

const setUserId = (userId): void => {
  if (!shouldTrackAnalytics()) { return }

  // make sure this is unique
  if (userId) {
    Amplitude.setUserIdAsync(userId)
  }
}

const setUserProperties = ({ ...userProperties }): void => {
  if (!shouldTrackAnalytics()) { return }

  Amplitude.setUserPropertiesAsync({ ...userProperties })
}

const clearUserProperties = (): void => {
  Amplitude.clearUserPropertiesAsync()
}

const logEvent = (eventName: string, eventProperties?: unknown): void => {
  if (!shouldTrackAnalytics()) { return }

  if (eventProperties) {
    Amplitude.logEventWithPropertiesAsync(eventName, eventProperties)
  } else {
    Amplitude.logEventAsync(eventName)
  }
}

const logScreenView = (screenName: string): void => {
  if (!shouldTrackAnalytics()) { return }

  logEvent(`VIEW_SCREEN_${screenName}`)
}

const events = {
  NOTIFICATIONS_ENABLE: `NOTIFICATIONS_ENABLE`,
  NOTIFICATIONS_SKIP: `NOTIFICATIONS_SKIP`,
  PUSH_NOTIFICATIONS_REGISTER: `PUSH_NOTIFICATIONS_REGISTER`,
  SETTINGS_GIVE_FEEDBACK: `SETTINGS_GIVE_FEEDBACK`,
  SETTINGS_RATE_APP: `SETTINGS_RATE_APP`,
  SETTINGS_VIEW_FAQS: `SETTINGS_VIEW_FAQS`,
}

export const AnalyticsManager = {
  clearUserProperties,
  events,
  initialise,
  logEvent,
  logScreenView,
  setUserId,
  setUserProperties,
}

export default AnalyticsManager
