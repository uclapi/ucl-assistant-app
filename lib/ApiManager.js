import axios from "axios";
import moment from "moment";
import { encode as btoa } from "base-64";
import {
  ROOMS_URL,
  WORKSPACES_URL,
  EQUIPMENT_URL,
  ROOMBOOKINGS_URL,
} from "../constants/API";

const roomsController = {
  search: async (token = null, query) => {
    if (query && query.length < 3) {
      return {};
    }
    try {
      const results = await axios.get(ROOMS_URL, {
        params: {
          query,
        },
        headers: {
          authorization: `Bearer ${token}`,
          "uclapi-roombookings-version": "1",
        },
      });
      if (!results.data.content.ok) {
        throw new Error(results.data.content.error);
      }
      return results.data.content.rooms;
    } catch (error) {
      throw error;
    }
  },
  getEquipment: async (token = null, { roomid, siteid }) => {
    if (!roomid || !siteid) {
      throw new Error("Must specify roomid and siteid");
    }
    try {
      const results = await axios.get(EQUIPMENT_URL, {
        params: {
          roomid,
          siteid,
        },
        headers: {
          authorization: `Bearer ${token}`,
        },
      });
      if (!results.data.content.ok) {
        throw new Error(results.data.content.error);
      }
      return results.data.content.equipment;
    } catch (error) {
      throw error;
    }
  },
  getBookings: async (
    token = null,
    { roomid, siteid, date = moment().format("YYYYMMDD") },
  ) => {
    if (!roomid || !siteid) {
      throw new Error("Must specify roomid and siteid");
    }
    try {
      const results = await axios.get(ROOMBOOKINGS_URL, {
        params: {
          roomid,
          siteid,
          date,
        },
        headers: {
          authorization: `Bearer ${token}`,
        },
      });
      if (!results.data.content.ok) {
        throw new Error(results.data.content.error);
      }
      return results.data.content.bookings;
    } catch (error) {
      throw error;
    }
  },
};

const workspacesApi = axios.create({
  baseURL: WORKSPACES_URL,
});

const workspacesController = {
  getLiveImage: async (token = null, { surveyId, mapId }) => {
    try {
      const result = await workspacesApi.get("/getliveimage/map.svg", {
        params: {
          survey_id: surveyId,
          map_id: mapId,
        },
        headers: {
          authorization: `Bearer ${token}`,
        },
        responseType: "arraybuffer",
      });
      const image = btoa(
        new Uint8Array(result.data).reduce(
          (data, byte) => data + String.fromCharCode(byte),
          "",
        ),
      );
      // there may be no need to base64 SVGs
      // but I have yet to test this extensively
      // https://css-tricks.com/probably-dont-base64-svg/
      return `data:image/svg+xml;base64,${image}`;
    } catch (error) {
      throw error;
    }
  },
};

export default {
  rooms: roomsController,
  workspaces: workspacesController,
};
