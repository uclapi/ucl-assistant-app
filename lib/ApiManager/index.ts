import Constants from 'expo-constants'
import PeopleController from './PeopleController'
import RoomsController from './RoomsController'
import TimetableController from './TimetableController'
import WorkspacesController from './WorkspacesController'

const clientId = Constants.manifest.extra?.uclapi?.clientId

export default {
  people: PeopleController,
  rooms: RoomsController,
  timetable: TimetableController,
  workspaces: WorkspacesController,
  clientId,
}
