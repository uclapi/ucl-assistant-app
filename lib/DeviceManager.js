import * as Device from 'expo-device'

const isRealDevice = () => Device.isRealDevice

export default {
  isRealDevice,
}
