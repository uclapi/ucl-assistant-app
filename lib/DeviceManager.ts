import * as Device from 'expo-device'
import * as Network from 'expo-network'
import { Platform } from 'react-native'

const isWeb = Platform.OS === `web`
const isAndroid = Platform.OS === `android`
const isiOS = Platform.OS === `ios`

const isRealDevice = Device.isDevice

const isConnectedToInternet = async (): Promise<boolean> => {
  if (isAndroid) {
    const isAirplaneMode = await Network.isAirplaneModeEnabledAsync()
    if (isAirplaneMode) {
      return false
    }
  } else {
    return false
  }

  const { isInternetReachable } = await Network.getNetworkStateAsync()
  return isInternetReachable
}

export default {
  isAndroid,
  isConnectedToInternet,
  isRealDevice,
  isWeb,
  isiOS,
}
