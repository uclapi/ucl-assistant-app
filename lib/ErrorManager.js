import Constants from 'expo-constants'
import * as Sentry from "sentry-expo"

const initialise = () => {
  if (!__DEV__) {
    const { extra: { sentry: { dsn } = {} } = {} } = Constants.manifest
    Sentry.init({
      dsn,
    })

    const { revisionId } = Constants.manifest
    if (typeof revisionId === `string` && revisionId.length > 0) {
      Sentry.setRelease(Constants.manifest.revisionId)
    }
  }
}

const captureError = (error) => {
  if (!__DEV__) {
    Sentry.captureException(error)
  } else {
    console.error(error)
  }
}

export default {
  captureError,
  initialise,
}
