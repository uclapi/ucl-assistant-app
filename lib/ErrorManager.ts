import Constants from 'expo-constants'
import * as Sentry from "sentry-expo"

const initialise = (): void => {
  if (!__DEV__) {
    const { extra: { sentry: { dsn = `` } = {} } = {}, revisionId } = Constants.manifest
    Sentry.init({
      dsn,
      ...(typeof revisionId === `string` && revisionId.length > 0
        ? { release: revisionId }
        : {}),
    })
  }
}

const addDetail = (details: any): void => {
  if (!__DEV__) {
    Sentry.Native.addBreadcrumb({
      ...details,
    })
  } else {
    console.info(`ErrorManager.addDetail`, details)
  }
}

const captureError = (error: Error, details?: any): void => {
  if (!__DEV__) {
    if (details) {
      Sentry.Native.withScope((scope) => {
        if (typeof details === `object`) {
          Object.entries(details).forEach(([attr, val]) => {
            scope.setExtra(attr, val)
          })
        } else {
          scope.setExtra(`details`, details)
        }
        Sentry.Native.captureException(error)
      })
    } else {
      Sentry.Native.captureException(error)
    }
  } else {
    console.error(`ErrorManager.captureError`, error, details)
  }
}

const setUser = (user): void => {
  if (!__DEV__ && user) {
    Sentry.Native.setUser(user)
  }
}

export default {
  addDetail,
  captureError,
  initialise,
  setUser,
}
