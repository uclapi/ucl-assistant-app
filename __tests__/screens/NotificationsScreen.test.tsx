/**
 * @jest-environment jsdom
 */
import "react-native"

import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import React from 'react'

import { PushNotificationsManager } from '../../lib'
import { NotificationsScreen } from "../../screens/NotificationsScreen"


Enzyme.configure({ adapter: new Adapter() })

describe(`NotificationsScreen`, () => {
  let wrapper
  const mockDispatch = jest.fn()
  const mockProps = {
    navigation: { dispatch: mockDispatch },
    token: `abc123`,
  }

  const mockRegisterForPushNotifications = jest.fn(() => Promise.resolve())
  PushNotificationsManager.registerForPushNotifications = (
    mockRegisterForPushNotifications
  )

  beforeEach(() => {
    wrapper = shallow(<NotificationsScreen {...mockProps} />)
  })

  it(`renders the NotificationsScreen`, () => {
    expect(wrapper).toMatchSnapshot()
  })

  it(
    `registers for notifications when onEnableNotifications runs`,
    async () => {
      await wrapper.instance().onEnableNotifications()

      expect(mockRegisterForPushNotifications).toHaveBeenCalledTimes(1)
      expect(mockRegisterForPushNotifications).toHaveBeenCalledWith(
        mockProps.token,
      )

      expect(mockDispatch).toHaveBeenCalledTimes(1)
    },
  )

  it(`does not register for notifications when onSkip runs`, async () => {
    await wrapper.instance().onSkip()

    expect(
      mockRegisterForPushNotifications,
    ).toHaveBeenCalledTimes(1) // i.e. not be called again
    expect(mockDispatch).toHaveBeenCalledTimes(2) // i.e. once more
  })
})
