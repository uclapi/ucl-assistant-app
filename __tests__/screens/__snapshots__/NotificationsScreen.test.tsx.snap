// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationsScreen renders the NotificationsScreen 1`] = `
<Page
  bottomColour="#fefefe"
  contentContainerStyle={Object {}}
  keyboardAvoidingViewStyle={Object {}}
  mainTabPage={false}
  onRefresh={[Function]}
  refreshEnabled={false}
  refreshing={false}
  safeAreaViewStyle={Object {}}
  style={Object {}}
  topColour="#fefefe"
>
  <View
    style={
      Object {
        "paddingBottom": 30,
        "paddingTop": 30,
      }
    }
  >
    <TitleText
      style={Object {}}
    >
      Be Notified
    </TitleText>
    <Image
      resizeMethod="scale"
      resizeMode="contain"
      source={1}
      style={
        Array [
          Object {
            "alignSelf": "stretch",
            "backgroundColor": "rgba(0,0,0,0)",
            "flex": 1,
            "height": undefined,
            "marginTop": 50,
            "width": undefined,
          },
          Object {
            "height": 150,
            "marginTop": 15,
          },
        ]
      }
    />
    <View
      style={
        Object {
          "marginBottom": 20,
          "marginTop": 20,
        }
      }
    >
      <BodyText
        style={Object {}}
      >
        Find out when there's a new feature on UCL Assistant,  when your timetable changes, or if there's an upcoming event that might  interest you.
      </BodyText>
    </View>
    <View
      style={
        Object {
          "marginBottom": 30,
        }
      }
    >
      <BodyText
        style={Object {}}
      >
        You can disable notifications later at any time.
      </BodyText>
    </View>
    <Button
      disabled={false}
      loading={false}
      onPress={[Function]}
      style={Object {}}
      testID="enable-notifications-button"
    >
      Enable Notifications
    </Button>
    <View
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "center",
          "marginBottom": 20,
          "marginTop": 20,
        }
      }
    >
      <Link
        containerStyle={Object {}}
        href=""
        onPress={[Function]}
        style={Object {}}
      >
        Skip
      </Link>
    </View>
  </View>
</Page>
`;
